lib/handlers/li.js

```diff
@@ -17,37 +17,43 @@ import {phrasing} from 'hast-util-phrasing'
  *   mdast node.
  */
 export function li(state, node) {
-  const head = node.children[0]
   /** @type {boolean | null} */
   let checked = null
   /** @type {Element | undefined} */
   let clone
 
-  // Check if this node starts with a checkbox.
-  if (head && head.type === 'element' && head.tagName === 'p') {
-    const checkbox = head.children[0]
+  // Check if this node starts with a checkbox (or a paragraph that starts with
+  // a checkbox), indicating the list item is a GFM TaskListItem.
+  let checkboxParent = node
+  let checkbox = node.children[0]
+  if (checkbox && checkbox.type === 'element' && checkbox.tagName === 'p') {
+    checkboxParent = checkbox
+    checkbox = checkbox.children[0]
+  }
 
-    if (
-      checkbox &&
-      checkbox.type === 'element' &&
-      checkbox.tagName === 'input' &&
-      checkbox.properties &&
-      (checkbox.properties.type === 'checkbox' ||
-        checkbox.properties.type === 'radio')
-    ) {
-      checked = Boolean(checkbox.properties.checked)
-      clone = {
-        ...node,
-        children: [
-          {...head, children: head.children.slice(1)},
-          ...node.children.slice(1)
-        ]
-      }
+  if (
+    checkbox &&
+    checkbox.type === 'element' &&
+    checkbox.tagName === 'input' &&
+    checkbox.properties &&
+    (checkbox.properties.type === 'checkbox' ||
+      checkbox.properties.type === 'radio')
+  ) {
+    checked = Boolean(checkbox.properties.checked)
+
+    clone = {...node, children: [...node.children]}
+    if (checkboxParent === node) {
+      clone.children.splice(0, 1)
+    } else {
+      clone.children.splice(0, 1, {
+        ...checkboxParent,
+        children: checkboxParent.children.slice(1)
+      })
     }
```

Comment:
a) some comments are useful
b) *operating* on the clone reads a bit complex, to me. Maybe that’s just me. What do you think about this?

```suggestion
    // The checkbox is the head of `li`:
    if (checkboxParent === node) {
      clone = {...node, children: node.children.slice(1)}
    }
    // The checkbox is in a paragraph (which is `checkboxParent`):
    else {
      clone = {
        ...node,
        children: [
          {...checkboxParent, children: checkboxParent.children.slice(1)},
          ...checkboxParent.children.slice(1)
        ]
      }
    }
```

Reply:
Ah! Yes, I definitely should have added some comments here. I tried both of these styles and liked on the one I committed slightly more, but I don’t have strong feelings. Will change.

Reply:
Note: will wait for feedback on the below item (about the checkbox with no text after it) before making the changes here, since there might be more to adjust depending on what the right output should be.

Reply:
Since I gather the main concern here is readability, I wound up extracting the code for finding and removing the checkbox into a separate function. I also wound up rewriting this recursively, which I think makes the logic much clearer to read (if you don’t think so, I can just roll back to f0acebfa6aa30dc75181e36e65c4a1d9137f9709, which is still a separate function but has your recommended implementation and some more comments and description).